WIP Metaflow on Airflow

valayDave opened the pull request:
## Metaflow on Airflow

## Current Support
- Datastore: `s3`
- Metadata: `service`
- compute: `@kubernetes`
- Internal decorators support : `@catch`,`@retry`, `@timeout`, `@environment`, `@schedule`, `@conda`, `@card`
- Parameters 

## Limitations
**No support for Foreachs (Coming In subsequent commits.)**

## Code Status :
- Unpolished. 
- Lots of TODOs. 

## Massive TODOs: 
- [x] Linear Workflow
- [x] Setting Identifiers correctly (task-id / run-id)
- [x] Configurable Parameters
- [x] Support for Other decorators like  `@catch`,`@retry`, `@timeout`, `@environment`,  `@conda`, `@card`
- [x] Branching 
- [x] Support for Max workers. 
- [x] Credential Management. 
- [ ] `IncludeFile` Testing 
- [x] Test with `@schedule` properly. 
- [x] Support `@project` decorator
- [x] K8s Secrets
- [x] Worker pools
- [ ] Foreach
- [ ] Test Cases
- [ ] Alert Decorator

valayDave committed:
commit 0
- unfinished code.

valayDave committed:
more tweeks.

valayDave committed:
First version of dynamic dag compiler.
- Not completely finished code
- Creates generic .py file a JSON that is parsed to create Airflow DAG.
- Currently only boiler plate to make a linear dag but doesn't execute anything.
-  Unfinished code.

valayDave committed:
some changes based on savin's comments.
- Added changes to task datastore for different reason : (todo) Decouple these
- Added comments to SFN for reference.
- Airflow DAG is no longer dependent on metaflow

valayDave committed:
Tweeks

valayDave committed:
Got linear workflows working on airflow.
- Still not feature complete as lots of args are still unfilled / lots of unknows
- minor tweek in eks to ensure airflow is k8s compatible.
- passing state around via xcom-push
- HACK : AWS keys are passed in a shady way. : Reverse this soon.

valayDave committed:
Fixed setting run-id / task-id to labels in k8s
- Fixed setting run-id has from cli macro
- added hashing macro to ensure that jinja template set the correct run-id to k8s labels
-

valayDave committed:
Added @schedule support + readme

valayDave committed:
Nit fix.

valayDave committed:
setting retry logic.

valayDave committed:
Fixed setting `task_id` :
- switch task-id from airflow job is to hash to "runid/stepname"
- refactor xcom setting variables
- added comments

valayDave committed:
Changed retry setting.

valayDave committed:
Fixed bug when using catch + retry

valayDave committed:
making changes sync with master

valayDave committed:
Merge branch 'master' into aft-mm

valayDave committed:
Fixing logs related change from master.

valayDave commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,402 @@
+# we can rename steps in a foreach with indexes (eg. `stepname-$index`) to create those steps.
+# Hence : Foreachs will require some special form of plumbing.
+# https://stackoverflow.com/questions/62962386/can-an-airflow-task-dynamically-generate-a-dag-at-runtime
+AIRFLOW_TASK_ID_TEMPLATE_VALUE = "arf-{{ [run_id, ti.task_id ] | task_id_creator  }}"
```
This is the macro we use to set `Task.id`

valayDave replied:
reasoning for it's choice is given in the code comments.

valayDave commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,402 @@
+    return hashlib.md5(my_value.encode("utf-8")).hexdigest()
+
+
+def task_id_creator(lst):
```
This is the templating function that helps create task-id from run-id and stepname.

valayDave replied:
This helps ensure that taskids dont change during retry attempts.

valayDave commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,532 @@
+        if schedule:
+            return schedule.schedule
+        # Airflow requires a scheduling arguement so keeping this
+        return "*/2 * * * *"
```
TODO : What should be defaults.

valayDave committed:
Removed todos + added some validation logic.

valayDave commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,528 @@
+
+    # Airflow run_ids are of the form : "manual__2022-03-15T01:26:41.186781+00:00"
+    # Such run-ids break the `metaflow.util.decompress_list`; this is why we hash the runid
+    run_id = "%s-$(echo -n {{ run_id }} | md5sum | awk '{print $1}')" % AIRFLOW_PREFIX
```
We try to retrieve md5-hash of run-id over here to fit it for airflow

valayDave committed:
Added support for Parameters.
- Supporting int, str, bool, float, JSONType

valayDave commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,566 @@
+                pass
+
+        # ! HACK : Remove Below Line; Only here temporarily.
+        env.update({k: v for k, v in os.environ.items() if "AWS" in k})
```
HACK HACK HACK. This is for very temporary testing purposes. Will be removed

valayDave committed:
Removed hacky line.

valayDave committed:
Added hacky line

valayDave commented on metaflow/plugins/airflow/compute/k8s.py:
```diff
@@ -0,0 +1,189 @@
+    return k8s_operator_args
+
+
+# todo : below code block only for reference. Remove it later.
```
The entire code block below this will be removed. It is only here for checking argument types.

valayDave committed:
Added support for Branching with Airflow
- remove `next` function in `AirflowTask`
- `AirflowTask`s has no knowledge of next tasks.
- removed todos + added some todos
- Graph construction on airflow side using graph_structure datastructure.
- graph_structure comes from`FlowGraph.output_steps()[1]`

valayDave committed:
fixing bugs in service account setting

valayDave committed:
2 changes :
- removed hacky line
- added support to directly throw dags in s3.

valayDave committed:
Adding exceptions for missing features.

valayDave committed:
Max workers and worker pool support.

valayDave committed:
@project, @schedule, default active dag support.
- Added a flag to allow setting dag as active on creation
- Airflow compatible schedule interval
- Project name fixes.

valayDave committed:
Bug fix in active defaults.

valayDave committed:
bug fix.

valayDave committed:
Bug fix.

valayDave committed:
Added secrets support.

valayDave committed:
Bug fixes WRT Kubernetes secrets + k8s deployments.
- Fixing some error messages.
- Added some comments.

savingoyal commented on metaflow/datastore/task_datastore.py:
```diff
@@ -264,36 +296,7 @@ def pickle_iter():
                     if isinstance(force_v4, bool)
                     else force_v4.get(name, False)
                 )
-                if do_v4:
```
This change doesn't seem to be doing anything specific here. Can we drop it?

valayDave replied:
Yes yes. This change was for something else but i think it got bundled along mistakenly.

savingoyal commented on metaflow/metaflow_config.py:
```diff
@@ -197,6 +197,11 @@ def from_conf(name, default=None):
+###
+# Airflow Configuration
+###
+AIRFLOW_STATE_MACHINE_PREFIX = from_conf("METAFLOW_AIRFLOW_STATE_MACHINE_PREFIX")
```
`METAFLOW_AIRFLOW_STATE_MACHINE_PREFIX ` - I don't think this environment variable is needed. Let's drop it.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+
+def resolve_state_machine_name(name):
+    def attach_prefix(name):
+        if AIRFLOW_STATE_MACHINE_PREFIX is not None:
```
There is no need for `STATE_MACHINE_PREFIX` for Airflow. It was introduced as part of a very specific customer-driven request for Step Functions which may not apply to Airflow.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+    is_project,
+    file_path=None,
+    worker_pool=None,
+    set_active=False,
```
What does `set_active` do here?

valayDave replied:
`set_active=True` means `is_paused_upon_creation=False`

savingoyal replied:
Let's replace it with `--is-paused-upon-creation` so that it is more intuitive to folks coming from the Airflow universe.

savingoyal commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,606 @@
+            # We need a separate unique ID for the special _parameters task
+            task_id_params = "%s-params" % self.task_id
+            # Export user-defined parameters into runtime environment
+            param_file = "".join(
```
Is this the only approach for handling parameters in Airflow? Don't they have first-class support for them?

valayDave committed:
Fixing bug fix in schedule.

savingoyal commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,606 @@
+            cmd = "if ! %s >/dev/null 2>/dev/null; then %s && %s; fi" % (
+                " ".join(exists),
+                export_params,
+                bash_capture_logs(" ".join(params)),
```
There shouldn't be any need for `bash_capture_logs` anymore.

valayDave replied:
Why so ?

savingoyal commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,606 @@
+        cmds.append(bash_capture_logs(" ".join(entrypoint + top_level + step)))
+        return " && ".join(cmds)
+
+    def _validate_workflow(self):
```
This check can be part of Airflow CLI itself. Currently, this check executes after you have packaged and stored the code in the datastore.

savingoyal commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,606 @@
+                        workflow,
+                    )
+
+            elif node.type == "foreach":
```
Since foreaches don't work right now, we can remove this conditional so that the next conditional can throw a pretty error.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+
+
+@airflow.command(help="Create an airflow workflow from this metaflow workflow")
+@click.argument("file_path", required=False)
```
Let's make this a required argument

valayDave committed:
Removing un-used code paths + code cleanup

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+    "if this option is not specified",
+)
+@click.option(
+    "--set-active",
```
let's name this to `--is-paused-upon-creation` so that it's more intuitive to folks coming from the Airflow world

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+        tags,
+        user_namespace,
+        max_workers,
+        False,
```
Need to enable support for `@project`

valayDave replied:
Project support is there. We switch names of the dag. I don't how else would I include it. I will clean up this bit from the code.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+    else:
+        if file_path.startswith("s3://"):
+            with S3() as s3:
+                s3.put(file_path, compiled_dag_file)
```
The S3 location for DAGs may not be the same S3 location for the datastore (and likely that the user has no way of authenticating to this bucket). After the demo - let's omit this functionality.

savingoyal commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,606 @@
+import json
+from metaflow.decorators import flow_decorators
+from metaflow.graph import FlowGraph, DAGNode
+from metaflow.plugins.cards.card_modules import chevron
```
Let's move `chevron` to `_vendor` - you make a new independent PR on top of master.

valayDave replied:
This is non-trivial as a lot of outside "metaflow" code uses chevron. An example is outside cards we / others may have published.

savingoyal commented on metaflow/plugins/airflow/airflow_compiler.py:
```diff
@@ -0,0 +1,606 @@
+)
+from . import airflow_utils as af_utils
+from .compute.k8s import create_k8s_args
+import metaflow.util as util
```
Minor nit - let's run `isort` on the imports before merging.

savingoyal commented on metaflow/plugins/airflow/README.md:
```diff
@@ -0,0 +1,19 @@
+# Airflow Integration 
```
We can remove this README for now

savingoyal commented on metaflow/plugins/airflow/af_deploy.py:
```diff
@@ -0,0 +1,9 @@
+import base64
```
The import statement should be inside `{{AIRFLOW_UTILS}}`

valayDave replied:
Which import statement? Do you mean `import base64` ? I think It's a rogue import. I will remove it.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+            raise MetaflowException(
+                "--name is not supported for @projects. " "Use --branch instead."
+            )
+        state_machine_name = attach_prefix(current.project_flow_name)
```
`state_machine` isn't a concept in Airflow. Let's replace it with `dag_name`.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+        return name
+
+    project = current.get("project_name")
+    if project:
```
Can you verify that this logic is consistent with the implementation in `argo-workflows` and `step-functions`? Are there any limitations to the length of the DAG name in Airflow?

valayDave replied:
I researched on dagname length limitations and couldn't find any explicitly. Will keep looking into it.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+VALID_NAME = re.compile("[^a-zA-Z0-9_\-\.]")
+
+
+def resolve_state_machine_name(name):
```
`resolve_dag_name`?

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+        state_machine_name = attach_prefix(current.project_flow_name)
+        is_project = True
+    else:
+        if name and VALID_NAME.search(name):
```
Does this hold true for Airflow?

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+    "--worker-pool",
+    default=None,
+    show_default=True,
+    help="Worker pool the for the airflow tasks.",
```
Nit: grammar.

savingoyal commented on metaflow/plugins/airflow/airflow_cli.py:
```diff
@@ -0,0 +1,172 @@
+    "--max-workers",
+    default=100,
+    show_default=True,
+    help="Maximum number of concurrent airflow tasks to run for the DAG. ",
```
Just `Maximum number of concurrent Airflow tasks.`

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,53 @@
+        ubf_context,
+        inputs,
+    ):
+        # todo : find out where the execution is taking place.
```
Can you expand on this TODO item?

valayDave replied:
XCOM's would work differently based on computing mediums. 

Example:  K8s Operators leverages XCOM via reading a JSON file present in the volume mounted by its sidecar container. 

So if we introduce Local/batch etc execution then we need to figure out how to pass the XCOM values.

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,53 @@
+        # Once figured where the execution is happening then we can do
+        # handle xcom push / pull differently
+        meta = {}
+        meta["airflow-execution"] = os.environ["METAFLOW_RUN_ID"]
```
What is `airflow-execution`? We already have the `METAFLOW_RUN_ID` by virtue of the fact that you need one to be able to access this metadata.

valayDave replied:
Spurious. Will Remove.

savingoyal commented on metaflow/plugins/airflow/airflow_decorator.py:
```diff
@@ -0,0 +1,53 @@
+            }
+        )
+
+    def task_finished(
```
We can omit overriding this function for now.

savingoyal commented on metaflow/plugins/airflow/compute/k8s.py:
```diff
@@ -0,0 +1,195 @@
+from datetime import timedelta
```
This can live entirely inside `airflow_compiler.py` given that we are only assembling a simple dictionary here.

savingoyal commented on metaflow/plugins/airflow/exceptions.py:
```diff
@@ -0,0 +1,20 @@
+from metaflow.exception import MetaflowException
+
+
+class AirflowNotPresent(MetaflowException):
```
Where is this exception thrown?

savingoyal commented on metaflow/plugins/airflow/exceptions.py:
```diff
@@ -0,0 +1,20 @@
+        super().__init__(msg)
+
+
+class NotSupportedException(MetaflowException):
```
This exception doesn't seem to be used right now anywhere.

valayDave replied:
Used when we detect foreachs.

savingoyal commented on metaflow/plugins/airflow/exceptions.py:
```diff
@@ -0,0 +1,20 @@
+        super().__init__(msg)
+
+
+class AirflowException(MetaflowException):
```
We can keep this exception in `airflow_compiler.py` and reduce file sprawl.

savingoyal commented on metaflow/plugins/airflow/plumbing/airflow_xcom_push.py:
```diff
@@ -0,0 +1,18 @@
+# How does xcom push work in K8s Operator : https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html#how-does-xcom-work
+import os
```
Minor - this module can simply reside in the parent directory

savingoyal replied:
Actually - this can wholly reside in `airflow_decorator.py`.

savingoyal commented on metaflow/plugins/airflow/plumbing/airflow_xcom_push.py:
```diff
@@ -0,0 +1,18 @@
+
+def safe_mkdir(dir):
+    try:
+        os.makedirs(dir)
```
What do we need to do to ensure that this directory is always writable? Can we throw a pretty error to the user if we detect that this isn't the case? For `argo-workflows`, we get around this limitation by mounting an emptyDir volume but it seems that isn't an option in Airflow.

valayDave replied:
Airflow communicates XCOM via a sidecar container that has `emptyDir` volume mounted to `/airflow/xcom`. There could be a condition where the side car has not started when the job container starts execution. So this condition ensures that directories are made safely. I don't think we need to raise any exception here.

savingoyal commented on metaflow/plugins/airflow/plumbing/set_parameters.py:
```diff
@@ -0,0 +1,24 @@
+
+
+def export_parameters(output_file):
+    input = json.loads(os.environ.get("METAFLOW_PARAMETERS", "{}"))
```
Do we need to handle parameters this way for Airflow? This specific approach was a workaround for all the limitations Step Functions was imposing on us. Perhaps there is a cleaner, more Airflow-native way of passing parameters?

valayDave replied:
Can you expand on what you mean by a more airflow native way? I use the `{{params}}` macro from airflow to pass the parameters to an environment variable. These are then used to set `METAFLOW_INIT_` env variable.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,477 @@
+AIRFLOW_TASK_ID_TEMPLATE_VALUE = "arf-{{ [run_id, ti.task_id ] | task_id_creator  }}"
+
+
+def sanitize_label_value(val):
```
JFYI - I will be removing this functionality in @kubernetes. Hashing this way has limited utility right now and it's better for us to understand the use cases rather than shipping a breaking change to the user later.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,477 @@
+    # This is a filter which creates a hash of the run_id/step_name string.
+    # Since run_ids in airflow are constants, they don't create an issue with the
+    #
+    return hashlib.md5("/".join(lst).encode("utf-8")).hexdigest()
```
Can you truncate the task-id to say 8 chars or so? They need to be unique within a step and it's a good guarantee that we won't run into collisions.

savingoyal replied:
Similar for the run-id that is generated. Since that needs to be globally unique - we can go with 12 chars.

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,477 @@
+    return hashlib.md5("/".join(lst).encode("utf-8")).hexdigest()
+
+
+def json_dump(val):
```
This function isn't really doing anything. Do we need it?

valayDave replied:
it is used as a `user_defined_filter` for airflow DAG

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,477 @@
+        },
+    }
+
+    metaflow_centric_args = {
```
Can we have a better name for this?

savingoyal commented on metaflow/plugins/airflow/airflow_utils.py:
```diff
@@ -0,0 +1,477 @@
+    return json.dumps(val)
+
+
+class AirflowDAGArgs(object):
```
This entire module can just live inside `af_deploy.py`?

valayDave committed:
Changes based on PR comments:
- removed airflow xcom push file , moved to decorator code
- removed prefix configuration
- nit fixes.

valayDave committed:
Moved exceptions to airflow compiler

valayDave committed:
Removing un-used args.

valayDave committed:
shortened length of run-id and task-id hashes.

valayDave committed:
changed `--is-active`->`--is-paused-upon-creation`
- dags are active by default.

valayDave committed:
Made file path required arguement.

valayDave committed:
Remove readme.

valayDave committed:
Removed usless code WRT project decorator.

valayDave committed:
Fixing mf log related bug
- No double logging on metaflow.

valayDave committed:
Added validations about:
- un-supported decorators
- foreach
Changed where validations are done to not save the package.

valayDave committed:
Refactored code .
- removed compute/k8s.py
- Moved k8s code to airflow_compiler.py
- ran isort to airflow_compiler.py

valayDave committed:
Fixed passing secrets with kubernetes.

valayDave committed:
3 changes:
- Removing s3 dep
- remove uesless import
- added `deployed_on` in dag file template

valayDave committed:
reverting change.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow-tests

valayDave committed:
ran black,

valayDave committed:
Removing un-used code / resolved-todos.

valayDave committed:
Merge branch 'airflow-tests' into airflow

valayDave committed:
variable renamemetaflow_specific_args

valayDave committed:
Option help bug fix.

valayDave committed:
fixed printing bug

valayDave committed:
Addressed comments.
